File: calculadora_cli.py
#!/usr/bin/env python3
"""
Calculadora Interativa de Terminal
Calculadora completa via Terminal (CLI) com cores ANSI, histórico e avaliação segura.
Execução: python calculadora_cli.py
"""

import sys
import math
import ast
import operator
from typing import Union, List

# Cores ANSI para visual no Terminal
AZUL = "\033[94m"
VERDE = "\033[92m"
AMARELO = "\033[93m"
VERMELHO = "\033[91m"
CIANO = "\033[96m"
NEGRITO = "\033[1m"
RESET = "\033[0m"


class SafeCalculator:
    """Avaliador de expressões matemáticas seguro usando AST (sem eval inseguro)."""

    OPERADORES = {
        ast.Add: operator.add,
        ast.Sub: operator.sub,
        ast.Mult: operator.mul,
        ast.Div: operator.truediv,
        ast.FloorDiv: operator.floordiv,
        ast.Mod: operator.mod,
        ast.Pow: operator.pow,
        ast.USub: operator.neg,
        ast.UAdd: operator.pos,
    }

    FUNCOES = {
        "sqrt": math.sqrt,
        "sin": math.sin,
        "cos": math.cos,
        "tan": math.tan,
        "log": math.log10,
        "ln": math.log,
        "abs": abs,
        "round": round,
        "fact": math.factorial,
    }

    CONSTANTES = {
        "pi": math.pi,
        "e": math.e,
        "tau": math.tau,
    }

    def __init__(self):
        self.historico: List[str] = []
        self.memoria: float = 0.0

    def avaliar(self, expressao: str) -> Union[int, float]:
        """Avalia com segurança uma expressão matemática."""
        arvore = ast.parse(expressao.strip(), mode="eval")
        return self._visitar(arvore.body)

    def _visitar(self, no):
        if isinstance(no, ast.Constant):  # Python 3.8+ (números)
            return no.value
        elif isinstance(no, ast.Name):
            nome = no.id.lower()
            if nome in self.CONSTANTES:
                return self.CONSTANTES[nome]
            elif nome == "ans" and self.historico:
                # Último resultado
                return float(self.historico[-1].split("=")[-1].strip())
            raise ValueError(f"Identificador desconhecido: '{no.id}'")
        elif isinstance(no, ast.BinOp):
            tipo_op = type(no.op)
            if tipo_op in self.OPERADORES:
                esquerdo = self._visitar(no.left)
                direito = self._visitar(no.right)
                if tipo_op == ast.Div and direito == 0:
                    raise ZeroDivisionError("Divisão por zero não é permitida.")
                return self.OPERADORES[tipo_op](esquerdo, direito)
            raise ValueError(f"Operador binário não suportado: {tipo_op.__name__}")
        elif isinstance(no, ast.UnaryOp):
            tipo_op = type(no.op)
            if tipo_op in self.OPERADORES:
                return self.OPERADORES[tipo_op](self._visitar(no.operand))
            raise ValueError("Operador unário não suportado.")
        elif isinstance(no, ast.Call):
            if isinstance(no.func, ast.Name):
                nome_func = no.func.id.lower()
                if nome_func in self.FUNCOES:
                    args = [self._visitar(arg) for arg in no.args]
                    return self.FUNCOES[nome_func](*args)
            raise ValueError("Chamada de função inválida ou desabilitada.")
        else:
            raise TypeError(f"Expressão não suportada: {type(no).__name__}")


def banner():
    print(f"{CIANO}{NEGRITO}=================================================={RESET}")
    print(f"{VERDE}{NEGRITO}           CALCULADORA PYTHON (CLI INTERATIVO)   {RESET}")
    print(f"{CIANO}{NEGRITO}=================================================={RESET}")
    print(f"Exemplos: 2 + 2, (10 * 5) / 2, sqrt(144), sin(pi / 2), 2 ** 8")
    print(f"Comandos especiais:")
    print(f"  {AMARELO}hist{RESET}  - Ver histórico de cálculos")
    print(f"  {AMARELO}m+{RESET}    - Adicionar valor atual à memória")
    print(f"  {AMARELO}mr{RESET}    - Recuperar valor da memória")
    print(f"  {AMARELO}mc{RESET}    - Limpar memória")
    print(f"  {AMARELO}ajuda{RESET} - Ajuda e funções disponíveis")
    print(f"  {VERMELHO}sair{RESET}  - Finalizar a calculadora")
    print(f"--------------------------------------------------\n")


def loop_calculadora():
    calc = SafeCalculator()
    banner()

    while True:
        try:
            entrada = input(f"{NEGRITO}calc > {RESET}").strip()
        except (KeyboardInterrupt, EOFError):
            print(f"\n{VERDE}Até logo!{RESET}")
            break

        if not entrada:
            continue

        cmd = entrada.lower()
        if cmd in ["sair", "exit", "quit", "q"]:
            print(f"{VERDE}Calculadora encerrada com sucesso.{RESET}")
            break
        elif cmd in ["hist", "historico"]:
            if not calc.historico:
                print(f"{AMARELO}Nenhum cálculo registrado ainda.{RESET}")
            else:
                print(f"{CIANO}--- Histórico de Cálculos ---{RESET}")
                for i, item in enumerate(calc.historico[-10:], 1):
                    print(f" {i}. {item}")
            continue
        elif cmd == "mc":
            calc.memoria = 0.0
            print(f"{AMARELO}Memória limpa: M = 0{RESET}")
            continue
        elif cmd == "mr":
            print(f"{VERDE}Memória (MR): {calc.memoria}{RESET}")
            continue
        elif cmd.startswith("m+"):
            try:
                val = float(cmd.split()[1]) if len(cmd.split()) > 1 else (float(calc.historico[-1].split("=")[-1].strip()) if calc.historico else 0.0)
                calc.memoria += val
                print(f"{VERDE}Memória atualizada: M = {calc.memoria}{RESET}")
            except Exception as e:
                print(f"{VERMELHO}Erro ao atualizar memória: {e}{RESET}")
            continue
        elif cmd in ["ajuda", "help", "?"]:
            print(f"{CIANO}Funções suportadas: sqrt(x), sin(x), cos(x), tan(x), log(x), ln(x), fact(x), abs(x)")
            print(f"Constantes: pi, e, tau | Operadores: +, -, *, /, //, %, **{RESET}")
            continue

        try:
            resultado = calc.avaliar(entrada)
            if isinstance(resultado, float):
                resultado = round(resultado, 8)
                if resultado.is_integer():
                    resultado = int(resultado)
            calc.historico.append(f"{entrada} = {resultado}")
            print(f"{VERDE}{NEGRITO}= {resultado}{RESET}")
        except ZeroDivisionError as err:
            print(f"{VERMELHO}Erro de Divisão: {err}{RESET}")
        except Exception as err:
            print(f"{VERMELHO}Erro: {err}{RESET}")


if __name__ == "__main__":
    loop_calculadora()

File: CalculadoraCLI.java
import java.util.*;

/**
 * CalculadoraCLI.java
 * Calculadora Interativa via Terminal CLI em Java Puro com Cores ANSI e Parser Matemático.
 * Execução: java CalculadoraCLI.java
 */
public class CalculadoraCLI {

    public static final String RESET = "\u001B[0m";
    public static final String VERDE = "\u001B[32m";
    public static final String AMARELO = "\u001B[33m";
    public static final String CIANO = "\u001B[36m";
    public static final String VERMELHO = "\u001B[31m";
    public static final String NEGRITO = "\u001B[1m";

    private double ans = 0.0;
    private double memoria = 0.0;
    private final List<String> historico = new ArrayList<>();
    private final int precisao = 6;

    public void iniciar() {
        Scanner scanner = new Scanner(System.in);
        System.out.println(CIANO + NEGRITO + "=================================================" + RESET);
        System.out.println(CIANO + NEGRITO + "          Calculadora Java (Terminal CLI)              " + RESET);
        System.out.println(CIANO + NEGRITO + "=================================================" + RESET);
        System.out.println("Comandos rápidos: " + AMARELO + "ajuda" + RESET + ", " + AMARELO + "hist" + RESET + ", " + AMARELO + "ans" + RESET + ", " + AMARELO + "m+" + RESET + ", " + AMARELO + "mr" + RESET + ", " + AMARELO + "sair" + RESET);
        System.out.println("Operações suportadas: +, -, *, /, ^, sqrt(x), sin(x), cos(x), abs(x)");
        System.out.println("-------------------------------------------------");

        while (true) {
            System.out.print(VERDE + NEGRITO + "calc> " + RESET);
            if (!scanner.hasNextLine()) break;
            String linha = scanner.nextLine().trim();

            if (linha.isEmpty()) continue;
            if (linha.equalsIgnoreCase("sair") || linha.equalsIgnoreCase("exit") || linha.equalsIgnoreCase("quit")) {
                System.out.println(CIANO + "Calculadora encerrada. Até logo!" + RESET);
                break;
            }

            processarComando(linha);
        }
    }

    private void processarComando(String cmd) {
        if (cmd.equalsIgnoreCase("ajuda") || cmd.equalsIgnoreCase("help")) {
            System.out.println(AMARELO + "--- Ajuda da Calculadora ---" + RESET);
            System.out.println(" • Expressões: 15 * 4 + sqrt(144) / 2");
            System.out.println(" • Variável ANS: armazena o último resultado calculado.");
            System.out.println(" • Memória: 'm+ 10', 'm- 5', 'mr' (ler memória), 'mc' (limpar)");
            System.out.println(" • Histórico: digite 'hist'");
            return;
        }

        if (cmd.equalsIgnoreCase("hist")) {
            System.out.println(AMARELO + "--- Histórico de Cálculos Recentes ---" + RESET);
            if (historico.isEmpty()) System.out.println(" Nenhum cálculo realizado ainda.");
            else for (int i = 0; i < historico.size(); i++) System.out.printf("  [%d] %s%n", i + 1, historico.get(i));
            return;
        }

        if (cmd.equalsIgnoreCase("ans")) {
            System.out.println(" ANS = " + VERDE + formatar(ans) + RESET);
            return;
        }

        if (cmd.equalsIgnoreCase("mr")) {
            System.out.println(" MEMÓRIA (MR) = " + VERDE + formatar(memoria) + RESET);
            return;
        }

        if (cmd.equalsIgnoreCase("mc")) {
            memoria = 0;
            System.out.println(AMARELO + "Memória zerada." + RESET);
            return;
        }

        if (cmd.startsWith("m+")) {
            String resto = cmd.substring(2).trim();
            double v = resto.isEmpty() ? ans : avaliar(resto);
            memoria += v;
            System.out.println(" Memória M+ " + VERDE + formatar(v) + RESET + " (Total: " + formatar(memoria) + ")");
            return;
        }

        try {
            double res = avaliar(cmd);
            ans = res;
            String linhaHist = String.format("%s = %s", cmd, formatar(res));
            historico.add(linhaHist);
            System.out.println(VERDE + NEGRITO + "  = " + formatar(res) + RESET);
        } catch (Exception e) {
            System.out.println(VERMELHO + "  Erro: " + e.getMessage() + RESET);
        }
    }

    public double avaliar(String expressao) {
        String exp = expressao.replaceAll("(?i)ans", String.valueOf(ans))
                              .replaceAll("(?i)pi", String.valueOf(Math.PI))
                              .replaceAll("(?i)e\\b", String.valueOf(Math.E))
                              .replaceAll("\\s+", "");
        return new Parser(exp).parse();
    }

    private String formatar(double val) {
        if (Double.isNaN(val) || Double.isInfinite(val)) return "Indefinido";
        if (val == (long) val) return String.valueOf((long) val);
        return String.format(Locale.US, "%." + precisao + "f", val).replaceAll("0+$", "").replaceAll("\\.$", "");
    }

    /**
     * Parser Matemático Seguro por Descida Recursiva (Sem eval).
     */
    private static class Parser {
        private final String str;
        private int pos = -1, ch;

        public Parser(String str) { this.str = str; nextChar(); }

        private void nextChar() { ch = (++pos < str.length()) ? str.charAt(pos) : -1; }

        private boolean eat(int charToEat) {
            while (ch == ' ') nextChar();
            if (ch == charToEat) { nextChar(); return true; }
            return false;
        }

        public double parse() {
            double x = parseExpression();
            if (pos < str.length()) throw new RuntimeException("Caractere inesperado: " + (char) ch);
            return x;
        }

        private double parseExpression() {
            double x = parseTerm();
            for (;;) {
                if (eat('+')) x += parseTerm();
                else if (eat('-')) x -= parseTerm();
                else return x;
            }
        }

        private double parseTerm() {
            double x = parseFactor();
            for (;;) {
                if (eat('*')) x *= parseFactor();
                else if (eat('/')) {
                    double divisor = parseFactor();
                    if (divisor == 0) throw new ArithmeticException("Divisão por zero.");
                    x /= divisor;
                } else if (eat('^')) {
                    x = Math.pow(x, parseFactor());
                } else return x;
            }
        }

        private double parseFactor() {
            if (eat('+')) return +parseFactor();
            if (eat('-')) return -parseFactor();

            double x;
            int startPos = this.pos;
            if (eat('(')) {
                x = parseExpression();
                if (!eat(')')) throw new RuntimeException("Parêntese ')' não fechado.");
            } else if ((ch >= '0' && ch <= '9') || ch == '.') {
                while ((ch >= '0' && ch <= '9') || ch == '.') nextChar();
                x = Double.parseDouble(str.substring(startPos, this.pos));
            } else if (ch >= 'a' && ch <= 'z') {
                while (ch >= 'a' && ch <= 'z') nextChar();
                String func = str.substring(startPos, this.pos);
                if (eat('(')) {
                    x = parseExpression();
                    if (!eat(')')) throw new RuntimeException("Parêntese ')' não fechado na função " + func);
                } else {
                    x = parseFactor();
                }
                switch (func.toLowerCase()) {
                    case "sqrt": x = Math.sqrt(x); break;
                    case "sin": x = Math.sin(Math.toRadians(x)); break;
                    case "cos": x = Math.cos(Math.toRadians(x)); break;
                    case "tan": x = Math.tan(Math.toRadians(x)); break;
                    case "abs": x = Math.abs(x); break;
                    case "log": x = Math.log10(x); break;
                    case "ln": x = Math.log(x); break;
                    default: throw new RuntimeException("Função desconhecida: " + func);
                }
            } else {
                throw new RuntimeException("Expressão inválida próxima a '" + (char) ch + "'");
            }
            return x;
        }
    }

    public static void main(String[] args) {
        new CalculadoraCLI().iniciar();
    }
}
